### ISSUE TITLE
Prometheus scrape occasionally fails with "server returned HTTP status 502 Bad Gateway" in Linkerd 2.7.1

### ISSUE THREAD

**Issue by ericsuhong** [linkerd/linkerd2#4274]
### What is the issue?

After upgrading Linkerd to 2.7.1 (from 2.7.0), Prometheus scraping occasionally fails on targets with "server returned HTTP status 502 Bad Gateway":

![image](https://user-images.githubusercontent.com/3857851/79683054-27dffb80-81dc-11ea-85f3-8f556435e2f1.png)

This error seems to be transient but consistent. As I keep refreshing Prometheus dashboard, these errors come and go across different pods consistently.

I see following errors in linkerd-proxy container from prometheus-server:
````
linkerd2_app_core::errors: Failed to proxy request: Service idled out after 60s
````

We do not have this problem in Linkerd 2.7.0.

We also do not have this problem when I run prometheus-server WITHOUT Linkerd injected.

### How can it be reproduced?

1. Install Linkerd 2.7.1
2. Install prometheus helm chart (10.4.0). Inject linkerd proxy into prometheus-server.
3. (We will use kured as an example pod).
Install Kured using a deployment yaml from here: https://github.com/weaveworks/kured.
Add following annotations to Kured deployment to make prometheus scrape from Kured:
````
annotations:
    linkerd.io/inject: disabled
    prometheus.io/scrape: "true"
    prometheus.io/port: "8080"
````
DO NOT inject linkerd proxy into kured.

4. Go to Prometheus dashboard by running these commands:
````
kubectl port-forward [prometheus pod name] 9090
localhost:9090
````
5. Go to Status -> Targets. Keep refreshing the page, and you will occasionally see Kured target being marked as "DOWN" with "**server returned HTTP status 502 Bad Gateway**"

### Logs, error output, etc

On linkerd-proxy pods, I see bunch of following errors:
````
WARN outbound:accept{peer.addr=10.240.1.154:42372}:source{target.addr=10.248.0.4:9100}: linkerd2_app_core::errors: Failed to proxy request: Service idled out after 60s
````
#### `linkerd check` output

```text
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
√ can initialize the client
√ can query the control plane API

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-identity
----------------
√ certificate config is valid
√ trust roots are using supported crypto algorithm
√ trust roots are within their validity period
√ trust roots are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
√ issuer cert is valid for at least 60 days
√ issuer cert is issued by the trust root

linkerd-api
-----------
√ control plane pods are ready
√ control plane self-check
√ [kubernetes] control plane can talk to Kubernetes
√ [prometheus] control plane can talk to Prometheus
√ tap api service is running

linkerd-version
---------------
√ can determine the latest version
√ cli is up-to-date

control-plane-version
---------------------
√ control plane is up-to-date
√ control plane and cli versions match

linkerd-ha-checks
-----------------
‼ pod injection disabled on kube-system
    kube-system namespace needs to have the label config.linkerd.io/admission-webhooks: disabled if HA mode is enabled
    see https://linkerd.io/checks/#l5d-injection-disabled for hints

Status check results are √
```

### Environment

- Kubernetes Version: 1.15.10
- Cluster Environment: aks-engine
- Host OS: Ubuntu 16.04
- Linkerd version: 2.17.1

**Comment by grampelberg** [comment]
Are you using resource requests and limits?

**Comment by ericsuhong** [comment]
Yes, we define only requests for prometheus via helm chart:

prometheus:
  server:
    resources:
      requests:
        cpu: 300m
        memory: 2000Mi

**Comment by grampelberg** [comment]
I'd bump the proxy requests to 1 cpu and see what happens.

**Comment by grampelberg** [comment]
Have you set `nf_conntrack_tcp_timeout_close_wait` on these btw?

**Comment by olix0r** [comment]
Are there any errors in prometheus's logs? Or does prometheus expose metrics indicating which specific scrape is failing? Basically trying to get at: are we sure `10.248.0.4:9100` is a live pod when this error is emitted?

**Comment by ericsuhong** [comment]
@grampelberg I increased CPU requests for both prometheus-server and linkerd-proxy, but it did not help. CPU usage for linkerd-proxy seems super low and normal (5~10 CPU millicores).

We only applied nf_conntrack_tcp_timeout_close_wait settings for nginx ingress controller pods.

**Comment by ericsuhong** [comment]
@olix0r Nothing shows up in prometheus logs. The only error logs messages I see is followings from linkerd-proxy container (from prometheus-server):
````
WARN outbound:accept{peer.addr=10.240.1.154:42372}:source{target.addr=10.248.0.4:9100}: linkerd2_app_core::errors: Failed to proxy request: Service idled out after 60s
````

Also, the pods are always live, because this is a test cluster with nothing major going on. I am consistently able to reproduce this issue, and as soon as I remove linkerd injection from prometheus, this issue goes away.

Note that problematic connections are not even meshes, since I see that failing targets do not run with linkerd injected.

**Comment by ericsuhong** [comment]
I enabled debug logs to linkerd-proxy, and here is the log file when "Service idled out after 60s" failure occurred twice:

[linkerd-proxy-failed-logs.txt](https://github.com/linkerd/linkerd2/files/4506938/linkerd-proxy-failed-logs.txt)

You can search using "Service idled out after 60s" in the log file.

**Comment by ericsuhong** [comment]
I also enabled "trace" level  and gathered logs for another pod failure here as well:
[linkerd-proxy-failed-trace-logs.txt](https://github.com/linkerd/linkerd2/files/4506956/linkerd-proxy-failed-trace-logs.txt)

**Comment by grampelberg** [comment]
@ericsuhong would you happen to have a very small yaml that replicates this? We'd love to give it a try from our side and see what the issue is.

**Comment by ericsuhong** [comment]
@grampelberg I provided a reprod steps (really just installing prometheus via helm and installing kured).

Can you see whether you can get a reprod from it?

**Comment by grampelberg** [comment]
I was kinda hoping for a YAML that I could apply. Worried you've got some configuration in your prometheus config that I'll be missing out on.

**Comment by ericsuhong** [comment]
I will try to get reprod yaml files, but do not have a time for it right now. Did you find anything interesting from attached logs?

**Comment by grampelberg** [comment]
No =( I think this is gonna require a tcpdump.

**Comment by ericsuhong** [comment]
@grampelberg I was able to get a reprod with a YAML. Can you please try the following YAML in a cluster with linkerd 2.7.1? After around 5 minutes (reprod takes some time), you will be able to see "**linkerd2_app_core::errors: Failed to proxy request: Service idled out after 60s**" errors coming from linkerd-proxy containers within prometheus server.

All components are installed in a namespace called "iam-system":
````
apiVersion: v1
kind: ServiceAccount
metadata:
  labels:
    component: "kube-state-metrics"
    app: prometheus
    release: iam-cluster-config
    chart: prometheus-10.4.0
    heritage: Helm
  name: iam-cluster-config-prometheus-kube-state-metrics
  namespace: iam-system
---
apiVersion: v1
kind: ServiceAccount
metadata:
  labels:
    component: "node-exporter"
    app: prometheus
    release: iam-cluster-config
    chart: prometheus-10.4.0
    heritage: Helm
  name: iam-cluster-config-prometheus-node-exporter
  namespace: iam-system
---
apiVersion: v1
kind: ServiceAccount
metadata:
  labels:
    component: "server"
    app: prometheus
    release: iam-cluster-config
    chart: prometheus-10.4.0
    heritage: Helm
  name: iam-cluster-config-prometheus-server
  namespace: iam-system
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: kured
  namespace: iam-system
---
apiVersion: v1
kind: ConfigMap
metadata:
  labels:
    component: "server"
    app: prometheus
    release: iam-cluster-config
    chart: prometheus-10.4.0
    heritage: Helm
  name: iam-cluster-config-prometheus-server
  namespace: iam-system
data:
  alerting_rules.yml: |
    {}
  alerts: |
    {}
  prometheus.yml: |
    global:
      evaluation_interval: 1m
      scrape_interval: 1m
      scrape_timeout: 10s
    rule_files:
    - /etc/config/recording_rules.yml
    - /etc/config/alerting_rules.yml
    - /etc/config/rules
    - /etc/config/alerts
    scrape_configs:
    - job_name: prometheus
      static_configs:
      - targets:
        - localhost:9090
    - bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
      job_name: kubernetes-apiservers
      kubernetes_sd_configs:
      - role: endpoints
      relabel_configs:
      - action: keep
        regex: default;kubernetes;https
        source_labels:
        - __meta_kubernetes_namespace
        - __meta_kubernetes_service_name
        - __meta_kubernetes_endpoint_port_name
      scheme: https
      tls_config:
        ca_file: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
        insecure_skip_verify: true
    - bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
      job_name: kubernetes-nodes
      kubernetes_sd_configs:
      - role: node
      relabel_configs:
      - action: labelmap
        regex: __meta_kubernetes_node_label_(.+)
      - replacement: kubernetes.default.svc:443
        target_label: __address__
      - regex: (.+)
        replacement: /api/v1/nodes/$1/proxy/metrics
        source_labels:
        - __meta_kubernetes_node_name
        target_label: __metrics_path__
      scheme: https
      tls_config:
        ca_file: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
        insecure_skip_verify: true
    - bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
      job_name: kubernetes-nodes-cadvisor
      kubernetes_sd_configs:
      - role: node
      relabel_configs:
      - action: labelmap
        regex: __meta_kubernetes_node_label_(.+)
      - replacement: kubernetes.default.svc:443
        target_label: __address__
      - regex: (.+)
        replacement: /api/v1/nodes/$1/proxy/metrics/cadvisor
        source_labels:
        - __meta_kubernetes_node_name
        target_label: __metrics_path__
      scheme: https
      tls_config:
        ca_file: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
        insecure_skip_verify: true
    - job_name: kubernetes-service-endpoints
      kubernetes_sd_configs:
      - role: endpoints
      relabel_configs:
      - action: keep
        regex: true
        source_labels:
        - __meta_kubernetes_service_annotation_prometheus_io_scrape
      - action: replace
        regex: (https?)
        source_labels:
        - __meta_kubernetes_service_annotation_prometheus_io_scheme
        target_label: __scheme__
      - action: replace
        regex: (.+)
        source_labels:
        - __meta_kubernetes_service_annotation_prometheus_io_path
        target_label: __metrics_path__
      - action: replace
        regex: ([^:]+)(?::\d+)?;(\d+)
        replacement: $1:$2
        source_labels:
        - __address__
        - __meta_kubernetes_service_annotation_prometheus_io_port
        target_label: __address__
      - action: labelmap
        regex: __meta_kubernetes_service_label_(.+)
      - action: replace
        source_labels:
        - __meta_kubernetes_namespace
        target_label: kubernetes_namespace
      - action: replace
        source_labels:
        - __meta_kubernetes_service_name
        target_label: kubernetes_name
      - action: replace
        source_labels:
        - __meta_kubernetes_pod_node_name
        target_label: kubernetes_node
    - job_name: kubernetes-service-endpoints-slow
      kubernetes_sd_configs:
      - role: endpoints
      relabel_configs:
      - action: keep
        regex: true
        source_labels:
        - __meta_kubernetes_service_annotation_prometheus_io_scrape_slow
      - action: replace
        regex: (https?)
        source_labels:
        - __meta_kubernetes_service_annotation_prometheus_io_scheme
        target_label: __scheme__
      - action: replace
        regex: (.+)
        source_labels:
        - __meta_kubernetes_service_annotation_prometheus_io_path
        target_label: __metrics_path__
      - action: replace
        regex: ([^:]+)(?::\d+)?;(\d+)
        replacement: $1:$2
        source_labels:
        - __address__
        - __meta_kubernetes_service_annotation_prometheus_io_port
        target_label: __address__
      - action: labelmap
        regex: __meta_kubernetes_service_label_(.+)
      - action: replace
        source_labels:
        - __meta_kubernetes_namespace
        target_label: kubernetes_namespace
      - action: replace
        source_labels:
        - __meta_kubernetes_service_name
        target_label: kubernetes_name
      - action: replace
        source_labels:
        - __meta_kubernetes_pod_node_name
        target_label: kubernetes_node
      scrape_interval: 5m
      scrape_timeout: 30s
    - honor_labels: true
      job_name: prometheus-pushgateway
      kubernetes_sd_configs:
      - role: service
      relabel_configs:
      - action: keep
        regex: pushgateway
        source_labels:
        - __meta_kubernetes_service_annotation_prometheus_io_probe
    - job_name: kubernetes-services
      kubernetes_sd_configs:
      - role: service
      metrics_path: /probe
      params:
        module:
        - http_2xx
      relabel_configs:
      - action: keep
        regex: true
        source_labels:
        - __meta_kubernetes_service_annotation_prometheus_io_probe
      - source_labels:
        - __address__
        target_label: __param_target
      - replacement: blackbox
        target_label: __address__
      - source_labels:
        - __param_target
        target_label: instance
      - action: labelmap
        regex: __meta_kubernetes_service_label_(.+)
      - source_labels:
        - __meta_kubernetes_namespace
        target_label: kubernetes_namespace
      - source_labels:
        - __meta_kubernetes_service_name
        target_label: kubernetes_name
    - job_name: kubernetes-pods
      kubernetes_sd_configs:
      - role: pod
      relabel_configs:
      - action: keep
        regex: true
        source_labels:
        - __meta_kubernetes_pod_annotation_prometheus_io_scrape
      - action: replace
        regex: (.+)
        source_labels:
        - __meta_kubernetes_pod_annotation_prometheus_io_path
        target_label: __metrics_path__
      - action: replace
        regex: ([^:]+)(?::\d+)?;(\d+)
        replacement: $1:$2
        source_labels:
        - __address__
        - __meta_kubernetes_pod_annotation_prometheus_io_port
        target_label: __address__
      - action: labelmap
        regex: __meta_kubernetes_pod_label_(.+)
      - action: replace
        source_labels:
        - __meta_kubernetes_namespace
        target_label: kubernetes_namespace
      - action: replace
        source_labels:
        - __meta_kubernetes_pod_name
        target_label: kubernetes_pod_name
    - job_name: kubernetes-pods-slow
      kubernetes_sd_configs:
      - role: pod
      relabel_configs:
      - action: keep
        regex: true
        source_labels:
        - __meta_kubernetes_pod_annotation_prometheus_io_scrape_slow
      - action: replace
        regex: (.+)
        source_labels:
        - __meta_kubernetes_pod_annotation_prometheus_io_path
        target_label: __metrics_path__
      - action: replace
        regex: ([^:]+)(?::\d+)?;(\d+)
        replacement: $1:$2
        source_labels:
        - __address__
        - __meta_kubernetes_pod_annotation_prometheus_io_port
        target_label: __address__
      - action: labelmap
        regex: __meta_kubernetes_pod_label_(.+)
      - action: replace
        source_labels:
        - __meta_kubernetes_namespace
        target_label: kubernetes_namespace
      - action: replace
        source_labels:
        - __meta_kubernetes_pod_name
        target_label: kubernetes_pod_name
      scrape_interval: 5m
      scrape_timeout: 30s
  recording_rules.yml: |
    {}
  rules: |
    {}
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRole
metadata:
  labels:
    component: "kube-state-metrics"
    app: prometheus
    release: iam-cluster-config
    chart: prometheus-10.4.0
    heritage: Helm
  name: iam-cluster-config-prometheus-kube-state-metrics
rules:
  - apiGroups:
      - ""
    resources:
      - namespaces
      - nodes
      - persistentvolumeclaims
      - pods
      - services
      - resourcequotas
      - replicationcontrollers
      - limitranges
      - persistentvolumeclaims
      - persistentvolumes
      - endpoints
      - secrets
      - configmaps
    verbs:
      - list
      - watch
  - apiGroups:
      - extensions
    resources:
      - daemonsets
      - deployments
      - ingresses
      - replicasets
    verbs:
      - list
      - watch
  - apiGroups:
      - apps
    resources:
      - daemonsets
      - deployments
      - statefulsets
      - replicasets
    verbs:
      - get
      - list
      - watch
  - apiGroups:
      - batch
    resources:
      - cronjobs
      - jobs
    verbs:
      - list
      - watch
  - apiGroups:
      - autoscaling
    resources:
      - horizontalpodautoscalers
    verbs:
      - list
      - watch
  - apiGroups:
      - policy
    resources:
      - poddisruptionbudgets
    verbs:
      - list
      - watch
  - apiGroups:
    - storage.k8s.io
    resources:
      - storageclasses
      - volumeattachments
    verbs:
      - list
      - watch
  - apiGroups:
      - certificates.k8s.io
    resources:
      - certificatesigningrequests
    verbs:
      - list
      - watch
  - apiGroups:
      - networking.k8s.io
    resources:
      - networkpolicies
    verbs:
      - list
      - watch
  - apiGroups:
      - admissionregistration.k8s.io
    resources:
      - validatingwebhookconfigurations
      - mutatingwebhookconfigurations
    verbs:
      - list
      - watch
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRole
metadata:
  labels:
    component: "server"
    app: prometheus
    release: iam-cluster-config
    chart: prometheus-10.4.0
    heritage: Helm
  name: iam-cluster-config-prometheus-server
rules:
  - apiGroups:
      - ""
    resources:
      - nodes
      - nodes/proxy
      - nodes/metrics
      - services
      - endpoints
      - pods
      - ingresses
      - configmaps
    verbs:
      - get
      - list
      - watch
  - apiGroups:
      - "extensions"
    resources:
      - ingresses/status
      - ingresses
    verbs:
      - get
      - list
      - watch
  - nonResourceURLs:
      - "/metrics"
    verbs:
      - get
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: kured
rules:
- apiGroups: [""]
  resources: ["nodes"]
  verbs:     ["get", "patch"]
- apiGroups: [""]
  resources: ["pods"]
  verbs:     ["list","delete","get"]
- apiGroups: ["apps"]
  resources: ["daemonsets"]
  verbs:     ["get"]
- apiGroups: [""]
  resources: ["pods/eviction"]
  verbs:     ["create"]
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  labels:
    component: "kube-state-metrics"
    app: prometheus
    release: iam-cluster-config
    chart: prometheus-10.4.0
    heritage: Helm
  name: iam-cluster-config-prometheus-kube-state-metrics
subjects:
  - kind: ServiceAccount
    name: iam-cluster-config-prometheus-kube-state-metrics
    namespace: iam-system
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: iam-cluster-config-prometheus-kube-state-metrics
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  labels:
    component: "server"
    app: prometheus
    release: iam-cluster-config
    chart: prometheus-10.4.0
    heritage: Helm
  name: iam-cluster-config-prometheus-server
subjects:
  - kind: ServiceAccount
    name: iam-cluster-config-prometheus-server
    namespace: iam-system
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: iam-cluster-config-prometheus-server
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: kured
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: kured
subjects:
- kind: ServiceAccount
  name: kured
  namespace: iam-system
---
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  namespace: iam-system
  name: kured
rules:
- apiGroups:     ["apps"]
  resources:     ["daemonsets"]
  resourceNames: ["kured"]
  verbs:         ["update"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  namespace: iam-system
  name: kured
subjects:
- kind: ServiceAccount
  namespace: iam-system
  name: kured
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: kured
---
apiVersion: v1
kind: Service
metadata:
  annotations:
    prometheus.io/port: "8080"
    prometheus.io/scrape: "true"
  labels:
    component: "kube-state-metrics"
    app: prometheus
    release: iam-cluster-config
    chart: prometheus-10.4.0
    heritage: Helm
  name: iam-cluster-config-prometheus-kube-state-metrics
  namespace: iam-system
spec:
  clusterIP: None
  ports:
    - name: http
      port: 80
      protocol: TCP
      targetPort: 8080
    - name: telemetry
      port: 81
      protocol: TCP
      targetPort: 8081
  selector:
    component: "kube-state-metrics"
    app: prometheus
    release: iam-cluster-config
  type: "ClusterIP"
---
apiVersion: v1
kind: Service
metadata:
  annotations:
    prometheus.io/port: "9100"
    prometheus.io/scrape: "true"
  labels:
    component: "node-exporter"
    app: prometheus
    release: iam-cluster-config
    chart: prometheus-10.4.0
    heritage: Helm
  name: iam-cluster-config-prometheus-node-exporter
  namespace: iam-system
spec:
  clusterIP: None
  ports:
    - name: metrics
      port: 9100
      protocol: TCP
      targetPort: 9100
  selector:
    component: "node-exporter"
    app: prometheus
    release: iam-cluster-config
  type: "ClusterIP"
---
apiVersion: v1
kind: Service
metadata:
  labels:
    component: "server"
    app: prometheus
    release: iam-cluster-config
    chart: prometheus-10.4.0
    heritage: Helm
  name: iam-cluster-config-prometheus-server
  namespace: iam-system
spec:
  ports:
    - name: http
      port: 80
      protocol: TCP
      targetPort: 9090
  selector:
    component: "server"
    app: prometheus
    release: iam-cluster-config
  sessionAffinity: None
  type: "ClusterIP"
---
apiVersion: apps/v1
kind: DaemonSet
metadata:
  labels:
    component: "node-exporter"
    app: prometheus
    release: iam-cluster-config
    chart: prometheus-10.4.0
    heritage: Helm
  name: iam-cluster-config-prometheus-node-exporter
  namespace: iam-system
spec:
  selector:
    matchLabels:
      component: "node-exporter"
      app: prometheus
      release: iam-cluster-config
  updateStrategy:
    type: RollingUpdate
  template:
    metadata:
      labels:
        component: "node-exporter"
        app: prometheus
        release: iam-cluster-config
        chart: prometheus-10.4.0
        heritage: Helm
    spec:
      serviceAccountName: iam-cluster-config-prometheus-node-exporter
      containers:
        - name: prometheus-node-exporter
          image: "prom/node-exporter:v0.18.1"
          imagePullPolicy: "IfNotPresent"
          args:
            - --path.procfs=/host/proc
            - --path.sysfs=/host/sys
          ports:
            - name: metrics
              containerPort: 9100
              hostPort: 9100
          resources:
            {}
          volumeMounts:
            - name: proc
              mountPath: /host/proc
              readOnly:  true
            - name: sys
              mountPath: /host/sys
              readOnly: true
      hostNetwork: true
      hostPID: true
      tolerations:
        - effect: NoSchedule
          key: node-role.kubernetes.io/master
          operator: Equal
          value: "true"
      volumes:
        - name: proc
          hostPath:
            path: /proc
        - name: sys
          hostPath:
            path: /sys
---
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: kured            # Must match `--ds-name`
  namespace: iam-system  # Must match `--ds-namespace`
spec:
  selector:
    matchLabels:
      name: kured
  updateStrategy:
   type: RollingUpdate
  template:
    metadata:
      annotations:
        linkerd.io/inject: disabled
        prometheus.io/scrape: "true"
        prometheus.io/port: "8080"
      labels:
        name: kured
    spec:
      serviceAccountName: kured
      tolerations:
        - effect: NoSchedule
          key: node-role.kubernetes.io/master
          operator: "Equal"
          value: "true"
      hostPID: true # Facilitate entering the host mount namespace via init
      restartPolicy: Always
      containers:
        - name: kured
          image: docker.io/weaveworks/kured:1.3.0
          imagePullPolicy: IfNotPresent
          securityContext:
            privileged: true # Give permission to nsenter /proc/1/ns/mnt
          env:
            # Pass in the name of the node on which this pod is scheduled
            # for use with drain/uncordon operations and lock acquisition
            - name: KURED_NODE_ID
              valueFrom:
                fieldRef:
                  fieldPath: spec.nodeName
          command:
            - /usr/bin/kured
            - --period=8h
            - --ds-namespace=iam-system
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    component: "kube-state-metrics"
    app: prometheus
    release: iam-cluster-config
    chart: prometheus-10.4.0
    heritage: Helm
  name: iam-cluster-config-prometheus-kube-state-metrics
  namespace: iam-system
spec:
  selector:
    matchLabels:
      component: "kube-state-metrics"
      app: prometheus
      release: iam-cluster-config
  replicas: 1
  template:
    metadata:
      labels:
        component: "kube-state-metrics"
        app: prometheus
        release: iam-cluster-config
        chart: prometheus-10.4.0
        heritage: Helm
    spec:
      serviceAccountName: iam-cluster-config-prometheus-kube-state-metrics
      containers:
        - name: prometheus-kube-state-metrics
          image: "quay.io/coreos/kube-state-metrics:v1.9.1"
          imagePullPolicy: "IfNotPresent"
          ports:
            - name: metrics
              containerPort: 8080
            - name: telemetry
              containerPort: 8081
          resources:
            {}
      securityContext:
        runAsNonRoot: true
        runAsUser: 65534
      tolerations:
        - effect: NoSchedule
          key: node-role.kubernetes.io/master
          operator: Equal
          value: "true"
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    component: "server"
    app: prometheus
    release: iam-cluster-config
    chart: prometheus-10.4.0
    heritage: Helm
  name: iam-cluster-config-prometheus-server
  namespace: iam-system
spec:
  selector:
    matchLabels:
      component: "server"
      app: prometheus
      release: iam-cluster-config
  replicas: 1
  template:
    metadata:
      annotations:
        linkerd.io/inject: enabled
      labels:
        component: "server"
        app: prometheus
        release: iam-cluster-config
        chart: prometheus-10.4.0
        heritage: Helm
    spec:
      serviceAccountName: iam-cluster-config-prometheus-server
      containers:
        - name: prometheus-server-configmap-reload
          image: "jimmidyson/configmap-reload:v0.3.0"
          imagePullPolicy: "IfNotPresent"
          args:
            - --volume-dir=/etc/config
            - --webhook-url=http://127.0.0.1:9090/-/reload
          resources:
            {}
          volumeMounts:
            - name: config-volume
              mountPath: /etc/config
              readOnly: true

        - name: prometheus-server
          image: "prom/prometheus:v2.15.2"
          imagePullPolicy: "IfNotPresent"
          args:
            - --storage.tsdb.retention.time=5d
            - --config.file=/etc/config/prometheus.yml
            - --storage.tsdb.path=/data
            - --web.console.libraries=/etc/prometheus/console_libraries
            - --web.console.templates=/etc/prometheus/consoles
            - --web.enable-lifecycle
          ports:
            - containerPort: 9090
          readinessProbe:
            httpGet:
              path: /-/ready
              port: 9090
            initialDelaySeconds: 30
            timeoutSeconds: 30
            failureThreshold: 3
            successThreshold: 1
          livenessProbe:
            httpGet:
              path: /-/healthy
              port: 9090
            initialDelaySeconds: 30
            timeoutSeconds: 30
            failureThreshold: 3
            successThreshold: 1
          resources:
            requests:
              cpu: 300m
              memory: 2000Mi
          volumeMounts:
            - name: config-volume
              mountPath: /etc/config
            - name: storage-volume
              mountPath: /data
              subPath: ""
      securityContext:
        fsGroup: 65534
        runAsGroup: 65534
        runAsNonRoot: true
        runAsUser: 65534
      tolerations:
        - effect: NoSchedule
          key: node-role.kubernetes.io/master
          operator: Equal
          value: "true"
      affinity:
        nodeAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
          - preference:
              matchExpressions:
              - key: kubernetes.io/role
                operator: In
                values:
                - master
            weight: 1
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
          - podAffinityTerm:
              labelSelector:
                matchExpressions:
                - key: app
                  operator: In
                  values:
                  - prometheus
              topologyKey: kubernetes.io/hostname
            weight: 1
      terminationGracePeriodSeconds: 300
      volumes:
        - name: config-volume
          configMap:
            name: iam-cluster-config-prometheus-server
        - name: storage-volume
          emptyDir:
            {}
---
````

**Comment by grampelberg** [comment]
@ericsuhong thanks! Have you observed this with anything other than kured?

**Comment by ericsuhong** [comment]
Yes, this issue occurs for all pods and endpoints actually. You don't actually need to install kured from above YAML to reproduce this issue.

**Comment by olix0r** [comment]
I think I have a handle on what's going on here... It seems to be a race between the 60s idle timeout and the 60s polling interval for prometheus. The gross workaround for this would be to modify the proxy's environment to include something like `LINKERD2_PROXY_OUTBOUND_ROUTER_MAX_IDLE_AGE: 70s`. I'll look into how we can avoid this.

**Comment by ericsuhong** [comment]
@olix0r In which situations does this issue have an impact? Does it impact regular web requests that take longer than 60 seconds to execute? Or does it only affect "every 60 seconds poll" type of workloads only?

**Comment by olix0r** [comment]
@ericsuhong The latter. The issue is that we drop clients that haven't received any requests in 60s; and prometheus manages to send a request at exactly the same time as the eviction. If it were a few milliseconds earlier or later, we wouldn't see the error.

I just built a proxy image that relaxes & jitters these timeouts to prevent this type of collision while we work out a better fix for this. For now, you can test this version by adding the `config.linkerd.io/proxy-version: ver-proxy-idle-jitter.0` annotation to your workload. I'm going to test this version over the weekend and, if it eliminates issues in my local repro setup, I'll put a PR up to be included in next week's edge release (and, subsequently, stable-2.8.0).

**Comment by ericsuhong** [comment]
@olix0r As far as I know, this is a regression introduced in 2.7.1 (we never had this issue on 2.7.0). Was there any changes in 2.7.1 that may have caused this issue to pop up?

Temporarily fixing this issue with a jitter is fine, but we probably won't upgrade until linkerd2 has a more permanent fix for the root cause.

**Comment by olix0r** [comment]
@ericsuhong Yes, 2.7.1 primarily included a change to how the proxy handles caching and buffering.

I don't expect that a proper fix will be that difficult now that we understand the issue.

**Comment by cpretzer** [comment]
nice find @olix0r 

@ericsuhong your repro case was perfect and extremely helpful, so thank you very much. In my test environment, I was able to prevent the error from occurring by setting the `LINKERD2_PROXY_OUTBOUND_ROUTER_MAX_IDLE_AGE`, which is a gross workaround (as @olix0r described) because it requires manually injecting the deployment like this:

```bash
kubectl get deploy iam-cluster-config-prometheus-server -oyaml | \
  linkerd inject --manual - | \
  kubectl apply -f -
```

Then set the environment variable:
```bash
kubectl set env -c linkerd-proxy deploy/iam-cluster-config-prometheus-server LINKERD2_PROXY_OUTBOUND_ROUTER_MAX_IDLE_AGE=70s
```

**Comment by olix0r** [comment]
I have a better fix available via
```
config.linkerd.io/proxy-version: gcr.io/linkerd-io/proxy:ver-idle-buffer-2020-05-04.1
```
 I'm currently testing this branch locally and [preparing a PR](https://github.com/linkerd/linkerd2-proxy/commit/8a862688094c94c6427817e876b643b07be3a3d6). The main goal of the change is to only check for idle timeouts after we have checked that no new requests are available. (Previously, idleness was checked before checking for new requests, allowing requests to fail with an idle error).

**Comment by olix0r** [comment]
The fix for this is on the most recent edge release and will be included in the upcoming `stable-2.8.0`. I'm going to close this, but please don't hesitate to reopen it if you observe this behavior on newer versions! And thanks again for the helpful report 😉
